component/org.wso2.carbon.identity.casque.authenticator/src/main/java/org/wso2/carbon/identity/casque/authenticator/authenticator/CasqueAuthenticator.java

```
@@ -0,0 +1,262 @@
+     */
+    private AuthenticatorFlowStatus start(HttpServletRequest request, HttpServletResponse response,
+                                          AuthenticationContext context) throws AuthenticationFailedException {
+
+        try {
+            String userName = request.getParameter(CasqueAuthenticatorConstants.USER_NAME);
+            if (StringUtils.isEmpty(userName)) {
+                throw new CasqueException("userName is null");
+            }
+
+            context.setProperty(CasqueAuthenticatorConstants.USER_NAME, userName);
+            String tokenId = getCasqueTokenId(userName);
+            String tokenIdPlusName = tokenId + userName;
+
+            context.setProperty(CasqueAuthenticatorConstants.RADIUS_STATE, null);
+            // Initial Access Request, fixed user,  token ID + username as the password
+            // Send request as Datagram packet
+            RadiusResponse radiusResponse = Radius.sendRequest("CASQUE SNR", tokenIdPlusName, null);
+            int radiusResponseType = radiusResponse.getType();
+
+            if (radiusResponseType == RadiusResponse.ACCESS_CHALLENGE) { // Got a challenge
+                context.setProperty(CasqueAuthenticatorConstants.RADIUS_STATE, radiusResponse.getState());
+                String challenge = radiusResponse.getChallenge();
+                String contextIdentifier = context.getContextIdentifier();
+                authPages.challengePage(response, contextIdentifier, challenge);
+                return AuthenticatorFlowStatus.INCOMPLETE;
+            }
+
+            clearProperties(context);
+
+            if (radiusResponseType == RadiusResponse.ACCESS_REJECT) {
+                throw new InvalidCredentialsException(
+                        " User authentication failed due to invalid credentials ", User.getUserFromUserName(userName));
+            }
+
+            throw new InvalidCredentialsException(" User authentication failed due to " + radiusResponse.getError(),
```

Do we really need to throw the InvalidCredentialsException here?